functions/geometry_vi.zsh

```diff
@@ -0,0 +1,25 @@
+# A geometry plugin to display the ZLE vi-mode state.
+
+update_rprompt() {
+    RPROMPT=${${RPROMPT/\[NORMAL\]/}/\[INSERT\]/}
+    RPROMPT="[$1]$RPROMPT"
+}
+
+update_prompt() {
+    PROMPT=${${PROMPT/\[NORMAL\]/}/\[INSERT\]/}
+    PROMPT="[$1]$PROMPT"
+}
+
+function zle-keymap-select {
+    keymap=$([[ $KEYMAP == vicmd ]] && echo "NORMAL" || echo "INSERT")
+    (( ${GEOMETRY_RPROMPT[(i)geometry_vi]} <= ${#GEOMETRY_RPROMPT} )) && update_rprompt $keymap
+    (( ${GEOMETRY_PROMPT[(i)geometry_vi]} <= ${#GEOMETRY_PROMPT} )) && update_prompt $keymap
+    zle reset-prompt
+}
+
+zle -N zle-keymap-select
+
+geometry_vi() {
+    # Empty render function, these are handled by zle-keymap-select
+}
```

Comment:
How about having the render function check if `zle-keymap-select` is enabled, and if not, add it.

```zsh
geometry_vi() {
   zle -l geometry-zle-vi-keymap-select || zle -N $_
}
```

Reply:
This is also kinda interesting because I think you might be able to put the prompt index logic in here instead of the keymap? not sure...